Support different base URIs

joneshf opened the pull request:
See #4.

I'm not entirely satisfied with this. It feels like the better way forward is to find out how to make it so orphan instances don't pop up. But, it seems to do what we discussed.

@owickstrom thoughts?

joneshf committed:
Support different base URIs

Choosing to always have `/` at the start of the URI meant
we could only send requests to the local server serving the code.

If we remove that restriction,
we can allow the API writer to decide where to send requests.

owickstrom commented on src/Type/Trout/Client.purs:
```diff
@@ -27,17 +33,38 @@ import Type.Trout.ContentType.JSON (JSON)
 
 toAffjaxRequest :: RequestBuilder -> AffjaxRequest Unit
-toAffjaxRequest req = defaultRequest { url = "/" <> joinWith "/" req.path }
+toAffjaxRequest req = defaultRequest { url = joinWith "/" req.path }
```
Doesn't this change mean that types without `BaseURI` will be relative? What about a `baseUri :: Maybe String` field in the `RequestBuilder`, or just a `String` defaulting to `/`?

joneshf replied:
Good idea. Adding a field sounds good.

joneshf committed:
Handle cases without a BaseURI better

The last change made all APIs without a baseURI relative.
Rather, you should be forced to include a baseURI if you don't need it.

Instead, we optionally prepend a baseURI,
but leave the rest of the API alone.

joneshf commented:
@owickstrom what do you think?

owickstrom commented:
@joneshf Thanks for the reminder! Everything looks good, except that maybe the `BaseURI` should not assume that you always want to have just the scheme and authority parts, e.g. when you run the API locally on `http://localhost` and in production you have `https://example.com/api/the-cool-thing` as base URI. Maybe have `BaseURI` wrap an `AbsoluteURI`?

nsaunders commented:
Resolved in #16
